modules/memberships/class-jetpack-memberships.php

```diff
@@ -276,5 +287,44 @@ public static function get_blog_id() {
 	public static function get_connected_account_id() {
 		return get_option( self::$connected_account_id_option_name );
 	}
+
+	/**
+	 * Whether Memberships (aka Recurring Payments) are enabled.
+	 *
+	 * @return bool
+	 */
+	public static function is_enabled_jetpack_recurring_payments() {
+		// For WPCOM sites.
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM && function_exists( 'has_any_blog_stickers' ) ) {
+			$site_id = get_current_blog_id();
+			return has_any_blog_stickers( array( 'premium-plan', 'business-plan', 'ecommerce-plan' ), $site_id );
+		}
+
+		// For all Jetpack sites.
+		return Jetpack::is_active() && Jetpack_Plan::supports( 'recurring-payments' );
+	}
+
+	/**
+	 * Register the Recurring Payments Gutenberg block
+	 */
+	public function register_gutenberg_block() {
```

Comment:
The name (`register_gutenberg_block`) feels a little generic vs the enclosed logic that only deals with a particular block (recurring-payments). I presume this is likely going to evolve into a more general block registration handler (?).

That said, is the underlying logic here a little specific for a general memberships class (or what appears to be)? I prefer to decouple things like these early on if that's the case. Fear being, it can get cluttered with specific block type references sooner than later, and harder to refactor later (or nobody bothers to do so). I do see the class is only a few months old though, so it probably makes sense for `v1`.

Just curious on perspective.

p.s. I guess there is the option of moving "block registration" to a trait as an intermediate solution.

Reply:
> The name (`register_gutenberg_block`) feels a little generic vs the enclosed logic that only deals with a particular block (recurring-payments).

Hmm, I was thinking of this method to mean "register this Jetpack module's Gutenberg block". I didn't really want to do a method named `register_xyz_gutenberg_block` inside of a class named `XYZ`, that seemed to redundant. (Okay, in this case, it would probably be `Jetpack_Memberships::register_recurring_payments_gutenberg_block()` rather than `Jetpack_Memberships::register_memberships_gutenberg_block()`, but I'd argue that  `memberships` and `recurring_payments` are mostly used synonymously in this file (and that the fact that two names exist for the same thing is mostly circumstantial).

> I presume this is likely going to evolve into a more general block registration handler (?).

I didn't intend to evolve it into somehting more general, TBH. The logic driving whether a block should be registered, or whether it should be set as unavailable, and for what reason, is something fairly specific to each individual block.

> That said, is the underlying logic here a little specific for a general memberships class (or what appears to be)? I prefer to decouple things like these early on if that's the case. Fear being, it can get cluttered with specific block type references sooner than later, and harder to refactor later (or nobody bothers to do so). I do see the class is only a few months old though, so it probably makes sense for `v1`. 
>
> Just curious on perspective - and the little I understand so far. :-)
> 
> p.s. I guess there is the option of moving "block registration" to a trait as an intermediate solution.

I agree in principle; but as stated above, I'm not sure we can make things a lot more homogenous here. We have fairly indidivual Jetpack modules (a somewhat legacy concept) each of which does things quite differently, and we often have to find the best place to register the corresponding block (or set their (un)availability).

I've so far considered the `jetpack_register_block()` and `Jetpack_Gutenberg::set_extension_unavailable`  our interface; modules can essentially use them at whatever point into their lifecycle (as long as it's soon enough for the consumer -- mostly the REST API -- to get accurate information). I guess for the purpose of the required plan check, we could require each module to provide information about the minimum plan it nees, and evaluate that information and register/set availability based on that. 

But this is complicated by the heterogenity of modules themselves. Note that not all modules come with a block; and not all blocks are tied to a module. Furthermore, a lot of modules exist both on JP and WP.com, but are implemented differently in both places.

And finally, there are other criteria besides required plan that can determine whether or not to show a block as available.

(That turned out a bit verbose :sweat_smile: I hope I didn't squash that -- commendable! -- idea of better consistency. If I'm beside the point, or if you had any other idea on how to do this, please let me know!)

Reply:
Well, this was an excellent overview writeup of the problem space to explain the _why's_ of the current setup. :-)

Would some of it deserve a place [in the docs?](https://github.com/Automattic/jetpack/tree/master/extensions#readme)

Reply:
Verbose is good. Thanks for the extra clarifications! :-)

I think the "modules" folder is a bit of a scary place for one to be (glad to see you mentioning "legacy" above)!

I can see this is following along how simple-payments is registering blocks. What I find missing I think is a contract binding these on common characteristics, or some abstraction of the things they do in common (which they apparently do - registering blocks for one). I believe you implied that in your answer above.

> if you had any other idea on how to do this, please let me know!

I don't, concretely, but I can throw some likely obvious ideas. 

Modules seems to be an all-encompassing folder here - so I guess I would start with that - add some structure to the things we know (in this case, "modules that register blocks"). I think I would approach this along these lines: 

Centralise the common logic elsewhere - provide a "yet another module" to handle block registration. I think I'd be happy if a module that registers a block defines some `options` for that block and only registers an action that takes care of the rest e.g.

1. define block options in the respective module (to be worked out in form, obviously):

```
class Simple_Payments {
    ...
    const OPTIONS = array( name => "recurring-payments", ... );
```

2. pass a callback to the action hooks:

```
class Simple_Payments {
    ...
    private function register_init_hooks() {
        add_action( 'jetpack_register_gutenberg_extension', Fancy_Module_Name::init( self::OPTIONS ) );
    }
```

3. with `Fancy_Module_Name` handling block registration. Simple, and presumably of the form:

```
class Fancy_Module_Name {
    static function init( $options ) {
        return function() use ( $options ) {
            // do stuff to register block with passed options/config
        };
    }
}
```

Alternatively, I guess the more OO way - a parent class that these modules can extend from, contract block registration functions via an interface, a trait that these modules can mix in, etc. Just stating the obvious perhaps - but something along these lines! (assuming my naive understanding of any underlying complexities)

Reply:
and my understanding is indeed fairly naive right now! I can hardly even tell how tightly-coupled the terms "membership" and "recurring-payments" are for one.